azure-application-insights-spring-boot-starter/src/main/java/com/microsoft/applicationinsights/boot/ApplicationInsightsWebMvcAutoConfiguration.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.boot;
+
+import com.microsoft.applicationinsights.TelemetryConfiguration;
+import com.microsoft.applicationinsights.internal.quickpulse.QuickPulse;
+import com.microsoft.applicationinsights.web.internal.WebRequestTrackingFilter;
+import com.microsoft.applicationinsights.web.spring.internal.InterceptorRegistry;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Import;
+import org.springframework.core.Ordered;
+
+@Configuration
+@Import(InterceptorRegistry.class)
+@ConditionalOnBean(TelemetryConfiguration.class)
+@ConditionalOnWebApplication
+@AutoConfigureAfter(ApplicationInsightsTelemetryAutoConfiguration.class)
+public class ApplicationInsightsWebMvcAutoConfiguration {
+
+    @Bean
+    public FilterRegistrationBean webRequestTrackingFilterRegistrationBean(WebRequestTrackingFilter webRequestTrackingFilter) {
+        FilterRegistrationBean registration = new FilterRegistrationBean();
+        registration.setFilter(webRequestTrackingFilter);
+        registration.addUrlPatterns("/*");
+        registration.setName("webRequestTrackingFilter");
+        registration.setOrder(Ordered.HIGHEST_PRECEDENCE + 10);
```

Comment:
Why do we need HIGHEST_PRECEDENCE+10. All the AI WebRequestTrackingFilter needs is that it should be first in the Filter Chain. Are there any spring framework filters that run even before the highest precedence? Also a question - is filter the best way to intercept the request execution time and other details? We have noticed an issue with Async request which returns a future and starts execution in different thread. In that case AI behaves weirdly. #444

Reply:
+10 just gives the user the ability to have some interceptor before ours. It may be useful in some cases for example when user want to log request or to measure overhead given by this filter. For instance we have filter called `FinalCatchFilter` which is the very first filter and logs any exception that was not handled in the chain.

Reply:
Okay, however my question was more centered around by providing this kind of facility would it mess up the actual numbers of request execution time? If that is the case we should be very careful with this. I would probably say no to it if it messes up the metrics.

Reply:
In general no one cares of order which is `Ordered.LOWEST_PRECEDENCE` by default. I see only two spring filters that is higher than `WebRequestTrackingFilter` -`MetricsFilter` and `OrderedCharacterEncodingFilter`, both of them are `Ordered.HIGHEST_PRECEDENCE` so you can't be sure that even with highest order yours win, also order you your concrete filter becomes non deterministic, which is not good thing.

If user specifies order that he is most probably know what is he doing and leaving some buffer is fine.

Preferable I would not touch it unless someone will give use case where this should be changed.

Reply:
@gavlyukovskiy I am not saying that we should provide the handle for user to tweak the filter. I only want to ensure that AI filter gets the highest precedence whenever possible but as you said it can be non deterministic, keeping the highest precedence + 10 should be a good enough guess in most cases.